assembly/src/todo/mod.rs

```diff
@@ -0,0 +1,432 @@
+use core::borrow::Borrow;
+
+use vm_core::{code_blocks::CodeBlock, CodeBlockTable, Felt, FieldElement, Operation, Program};
+
+use crate::{
+    parsers::{self, Instruction, Node},
+    procedures::Procedure,
+    ModuleProvider, ProcedureAst, ProcedureId, ProgramAst,
+};
+
+pub struct Assembler {
+    module_provider: Box<dyn ModuleProvider>,
+    cb_table: CodeBlockTable,
+    locals: Vec<Procedure>,
+    blocks: Vec<Vec<CodeBlock>>,
+    spans: Vec<Vec<Operation>>,
+}
+
+impl Assembler {
+    pub fn new() -> Self {
+        Self {
+            module_provider: Box::new(()),
+            cb_table: CodeBlockTable::default(),
+            locals: Vec::new(),
+            blocks: Vec::new(),
+            spans: Vec::new(),
+        }
+    }
+
+    pub fn with_cb_table(mut self, cb_table: CodeBlockTable) -> Self {
+        self.cb_table = cb_table;
+        self
+    }
+
+    pub fn with_module_provider<P>(mut self, provider: P) -> Self
+    where
+        P: ModuleProvider + 'static,
+    {
+        self.module_provider = Box::new(provider);
+        self
+    }
+
+    pub fn into_inner(self) -> CodeBlockTable {
+        self.cb_table
+    }
+
+    pub fn compile<S>(&mut self, source: S) -> Program
+    where
+        S: AsRef<str>,
+    {
+        let source = source.as_ref();
+        let ast = ProgramAst::try_from(source).unwrap();
+
+        self.compile_ast(&ast)
+    }
+
+    pub fn compile_ast(&mut self, ast: &ProgramAst) -> Program {
+        let ProgramAst { local_procs, body } = ast;
+
+        self.locals.clear();
+        self.blocks.clear();
+        self.spans.clear();
+
+        let root = self.compile_body(body.iter());
+
+        Program::new(root)
+    }
+
+    fn start_span<I, T>(&mut self, iter: &I)
+    where
+        I: Iterator<Item = T>,
+    {
+        self.end_span();
+        self.spans.push(vec_with_iter_capacity(iter));
+    }
+
+    fn append_to_span(&mut self, operation: Operation) {
+        if let Some(span) = self.spans.last_mut() {
+            span.push(operation);
+        }
+    }
+
+    fn end_span(&mut self) {
+        self.spans
+            .pop()
+            .filter(|v| !v.is_empty())
+            .map(CodeBlock::new_span)
+            .map(|span| self.blocks.last_mut().map(|block| block.push(span)));
+    }
+
+    fn start_block<I, T>(&mut self, iter: &I)
+    where
+        I: Iterator<Item = T>,
+    {
+        self.blocks.push(vec_with_iter_capacity(iter));
+    }
+
+    fn pop_block(&mut self) -> Option<CodeBlock> {
+        self.blocks
+            .pop()
+            .filter(|v| !v.is_empty())
+            .map(parsers::combine_blocks)
+    }
+
+    fn append_block(&mut self, block: CodeBlock) {
+        self.blocks.last_mut().map(|blocks| blocks.push(block));
+    }
+
+    fn compile_body<A, N>(&mut self, ast: A) -> CodeBlock
+    where
+        A: Iterator<Item = N>,
+        N: Borrow<Node>,
+    {
+        self.start_block(&ast);
+        self.start_span(&ast);
```

Comment:
Instead of putting `spans` and `blocks` as member variables into the assembler, I would make the local variables of this function.

Reply:
SpanBuilder created